pkg/operation/botanist/infrastructure.go

```diff
@@ -16,147 +16,81 @@ package botanist
-// for a successful reconciliation of an infrastructure resource.
-const InfrastructureDefaultTimeout = 5 * time.Minute
+// DefaultInfrastructure creates the default deployer for the Infrastructure custom resource.
+func (b *Botanist) DefaultInfrastructure(seedClient client.Client) shoot.Infrastructure {
+	return infrastructure.New(
+		b.Logger,
+		seedClient,
+		&infrastructure.Values{
+			Namespace:                               b.Shoot.SeedNamespace,
+			Name:                                    b.Shoot.Info.Name,
+			Type:                                    b.Shoot.Info.Spec.Provider.Type,
+			ProviderConfig:                          b.Shoot.Info.Spec.Provider.InfrastructureConfig,
+			Region:                                  b.Shoot.Info.Spec.Region,
+			IsInCreationPhase:                       b.Shoot.Info.Status.LastOperation != nil && b.Shoot.Info.Status.LastOperation.Type == gardencorev1beta1.LastOperationTypeCreate,
+			IsWakingUp:                              !gardencorev1beta1helper.HibernationIsEnabled(b.Shoot.Info) && b.Shoot.Info.Status.IsHibernated,
+			IsInRestorePhaseOfControlPlaneMigration: b.isRestorePhase(),
+			DeploymentRequested:                     controllerutils.HasTask(b.Shoot.Info.Annotations, common.ShootTaskDeployInfrastructure),
+		},
+		infrastructure.DefaultInterval,
+		infrastructure.DefaultSevereThreshold,
+		infrastructure.DefaultTimeout,
+	)
+}
 
-// DeployInfrastructure creates the `Infrastructure` extension resource in the shoot namespace in the seed
-// cluster. Gardener waits until an external controller did reconcile the cluster successfully.
+// DeployInfrastructure deploys the Infrastructure custom resource and triggers the restore operation in case
+// the Shoot is in the restore phase of the control plane migration.
 func (b *Botanist) DeployInfrastructure(ctx context.Context) error {
-	var (
-		operation                      = v1beta1constants.GardenerOperationReconcile
-		lastOperation                  = b.Shoot.Info.Status.LastOperation
-		creationPhase                  = lastOperation != nil && lastOperation.Type == gardencorev1beta1.LastOperationTypeCreate
-		shootIsWakingUp                = !gardencorev1beta1helper.HibernationIsEnabled(b.Shoot.Info) && b.Shoot.Info.Status.IsHibernated
-		restorePhase                   = b.isRestorePhase()
-		requestInfrastructureOperation = creationPhase || shootIsWakingUp || restorePhase || controllerutils.HasTask(b.Shoot.Info.Annotations, common.ShootTaskDeployInfrastructure)
-		infrastructure                 = &extensionsv1alpha1.Infrastructure{
+	b.Shoot.Components.Extensions.Infrastructure.SetSSHPublicKey(b.Secrets[v1beta1constants.SecretNameSSHKeyPair].Data[secrets.DataKeySSHAuthorizedKeys])
+
+	if err := b.Shoot.Components.Extensions.Infrastructure.Deploy(ctx); err != nil {
+		return err
+	}
+
+	if b.isRestorePhase() {
+		return b.restoreExtensionObject(ctx, b.K8sSeedClient.Client(), &extensionsv1alpha1.Infrastructure{
```

Comment:
I think this similarly to the Network will throw a nil pointer, because inside the restore we are trying to access some fields from the Infrastructure obj. May be we can temporally add  `GET` of the Infra object, until @plkokanov 
 change is merged.

Reply:
@swilen-iwanow hm, how does it work for `Network`s currently? There is no `GET` call either, or am I missing something?
https://github.com/gardener/gardener/blob/5773c96710adabe61574f39428ab2a6d43645a1f/pkg/operation/botanist/network.go#L55-L62

Reply:
No. My point was that with the change of the `Network`s to use a component approach and passing an empty object to the `restoreExtensionObject`, instead of the created object like it used to be, the restore scenario for the `Network`s is not working anymore. With the Infra it will be the same. What I'm saying is that we can add a quick fix by getting the object inside the `restoreExtensionObject` func, which could be done in a separate PR.

Reply:
OK, understood, thanks. Can you open such a PR for the `Network`? I'm happy to incorporate the changes in this PR as well. Or, you wait until this PR is merged and do it together at once for both `Network` and `Infrastructure`.

Reply:
I can wait until this PR is merged. If #2511 is merged as well I will just implement the new interfaces that are supposed to fix this